File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
  const dr = [0, 0, -1, 1]; // 상, 하, 좌, 우
  const dc = [1, -1, 0, 0]; // 상, 하, 좌, 우
  const n = maps.length;
  const m = maps[0].length;
  const visited = Array.from({ length: n }, () => Array(m).fill(false));

  const queue = [];
  queue.push([0, 0, 1]); // x좌표, y좌표, 거리

  while (queue.length > 0) {
    const [x, y, distance] = queue.shift();

    if (x === n - 1 && y === m - 1) return distance;

    for (let i = 0; i < 4; i++) {
      const nx = x + dr[i];
      const ny = y + dc[i];

      if (nx >= 0 && nx < n && ny >= 0 && ny < m) {
        if (maps[nx][ny] === 1 && !visited[nx][ny]) {
          visited[nx][ny] = true;
          queue.push([nx, ny, distance + 1]);
        }
      }
    }
  }
  return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    static int[][] visited;
    static int[] dr = {1, 0, -1, 0};
    static int[] dc = {0, 1, 0, -1};
    static int sum = 1;
    
    public int solution(int[][] maps) {
        visited = new int[maps.length][maps[0].length];
        bfs(maps);
        int row = maps.length - 1, column = maps[0].length - 1;
        if (visited[row][column] == 0) return -1;
        return visited[row][column];
    }
    
    public static void bfs(int[][] maps) {
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{0, 0});
        visited[0][0] = 1;
        
        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            
            for (int i = 0; i < 4; i++) {
                int nr = current[0] + dr[i];
                int nc = current[1] + dc[i];
                
                if (nr >= 0 && nc >= 0 && nr < visited.length && nc < visited[0].length) {
                    if (maps[nr][nc] == 1 && visited[nr][nc] == 0) {
                        queue.add(new int[]{nr, nc});
                        visited[nr][nc] = visited[current[0]][current[1]] + 1;
                    }
                }
            }
        }
        
    }
    
   
}